Objects/unionobject.c

```diff
@@ -237,9 +237,20 @@ dedup_and_flatten_args(PyObject* args)
         PyObject* i_element = PyTuple_GET_ITEM(args, i);
         for (Py_ssize_t j = i + 1; j < arg_length; j++) {
             PyObject* j_element = PyTuple_GET_ITEM(args, j);
-            if (i_element == j_element) {
-                is_duplicate = 1;
+            // RichCompare to also deduplicate GenericAlias types (slower)
+            if (Py_TYPE(i_element) == &Py_GenericAliasType &&
+                Py_TYPE(j_element) == &Py_GenericAliasType) 
+            {
+                if (PyObject_RichCompareBool(i_element, j_element, Py_EQ) == 1) {
```

Comment:
You need error checking here, in case the comparison fails.

Reply:
Sorry, I'm not very well-versed in the C parts of Python so this is a genuine question: are you referring to the `PyObject_RichCompareBool` comparison? 

If so, the [documentation for that](https://docs.python.org/3/c-api/object.html#c.PyObject_RichCompareBool), doesn't seem capable of raising an error or returning `NULL`; only `-1` (for error), `0` (false) or `1` (true).  But I can't picture a scenario where `-1` would arise since unions via `|` are only created by type objects and genericalias (which both have proper `__eq__` defined).

Reply:
Exactly, it returns -1 for errors, and who knows what errors might happen. It's an API function that is documented to return -1 for errors, so you *must* check for errors. If you don't, 5 years in the future something will make it return -1 and this code will do the wrong thing and the error will appear at a mysterious point much later. ALl you have to do is return an error from this function.

Reply:
Alright, I understand now, thanks for the explanation.  🙂

Reply:
The new error looks like this:

```python
>>> list[int] | list[int]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: Could not compare objects of type 'types.GenericAlias' and 'types.GenericAlias' at indexes 0 and 1 respectively. Their __eq__ methods may be invalid.
```